akka-actor/src/main/scala/akka/util/OptionVal.scala

```diff
@@ -0,0 +1,66 @@
+/**
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.util
+
+/**
+ * INTERNAL API
+ */
+private[akka] object OptionVal {
+
+  def apply[A](x: A): OptionVal[A] =
+    new OptionVal(x)
+
+  /**
+   * Represents non-existent values, `null` values.
+   */
+  val None = new OptionVal(null)
+}
+
+/**
+ * INTERNAL API
+ * Represents optional values similar to `scala.Option`, but
+ * as a value class to avoid allocations.
+ *
+ * Note that it should not be used in pattern matching to avoid
+ * allocations. That is also the reason why it's not a `case class`
+ * with `unapply`.
+ * See http://docs.scala-lang.org/overviews/core/value-classes.html#when-allocation-is-necessary
+ */
+private[akka] final class OptionVal[+A](val x: A) extends AnyVal {
```

Comment:
Awesome, finally :)
Methods on it look good.
Will raincheck if its the right name based extractio hmm, I think so